### PR TITLE
Dependency rework

### PR DESCRIPTION
**Pull request by stairman06**
Controversial! Allows dependencies with no MODIP metadata to exist. Somewhat related to #12.

Checklist:
- [X] Update spec
- [X] Update implementation guide
- [x] Provide full examples
- [x] Section providing full reasoning as to why this exists

### CODE REVIEW

**Comment by expeehaa** [comment]
I actually like the idea. This may have not been that clear from the discussions on discord, but since this PR lets the implementation handle a missing `src` field however it wants, it seems fine to me.

**Comment by stairman06** [comment]
Thanks. Leaving it up to implementation is the best way to go. If something like mod ids is supported to allow checking for dependencies, it will be very generic and not tied to specific modloaders (just a plain key/value map with some optional standardized IDs like for frameworks), and up to the implementation as well. But that is maybe out of scope for this PR.

File: format_implementing.md [hunk 1]
```diff
@@ -16,10 +16,11 @@ Framework IDs SHOULD be prefixed with `framework-`. This naming scheme makes it
...
 
-For the purposes of this section, the term "the host" refers to the service a launcher is using in order to download projects, such as Diluv or Modrinth.
-
-If a dependency has the `src` field, it means that at that specified URL is full metadata about this dependency. Launchers should check here first if it's specified. If the URL does not serve a suitable result, or there is no `src` field, launchers should then ask the host of the parent for a project matching the same ID. If at this point the launcher still hasn't found any metadata for the project, the launcher may choose to ask another host that it knows of for a project with the same ID, or it may choose to fail and warn the user.
+When looking at a dependency, the first thing to do is check for the `src` field. If the `src` field exists, then the mod has external MODIP metadata. Process the metadata for that dependency from the `src` field.
 
+If there is no `src` field, you have two options:
+1) Request information on this ID from hosts that your launcher knows of.
+2) Throw a warning to the user, suggesting that they need to download the project, but the launcher cannot automatically install it.
```
Comment (immibis) [inline format_implementing.md:23]: only *if* the project isn't already installed, right?

Comment (stairman06) [reply format_implementing.md:23]: Correct. Detecting if a project is installed is complicated and left mostly up to the implementation. If/when modids or another similar system is supported, then there will be _some_ guidelines on how to implement the system, but it's still up to the implementation.

**Comment by stairman06** [comment]
Reasoning and an example has been added.

File: format_implementing.md [hunk 2]
```diff
@@ -16,9 +16,34 @@ Framework IDs SHOULD be prefixed with `framework-`. This naming scheme makes it
...
+### Example
+
+In the below example, a launcher should display a message indicating that `non-modip-dependency` cannot be automatically downloaded.
+```json
+"dependencies": [
+  {
+    "id": "non-modip-dependency",
+    "name": "Non-MODIP Dependency",
+    "version": "1.0.0",
+    // notice: no `src` field
```
Comment (immibis) [inline format_implementing.md:41]: No mod ID or expected class files either?

File: format_spec.md [hunk 3]
```diff
@@ -324,11 +324,12 @@ This field MUST be either an Array or String. If this field is an array, it MUST
 
 ##### `src` (optional)
 
-This specifies where information about this dependency can be acquired, but it's not always required. If the metadata of the parent project is being served from a host, then this field is not required and clients can lookup the ID at the host. For example, if I request metadata about a project from Modrinth and there is a dependency, "my-dependency", then I should call Modrinth's API in order to find out more about that dependency.
+This specifies where information about this dependency can be acquired. This field is *not* required as MODIP metadata for everything cannot be provided. 
 
-But it some cases, that's either not possible or impractical, which is where the `src` field comes in. This field MUST contain either a String containing a URL that links to full metadata about the dependency, or it can be an object which fully complies with the MODIP project spec.
+If this field is provided, it MUST contain either a String containing a URL that links to full metadata about the dependency, or it can be an object which fully complies with the MODIP project spec.
```
Comment (immibis) [inline format_spec.md]: So any API defined in the future cannot be used here?

File: format_implementing.md (hunk 2 above)
Comment (stairman06) [reply format_implementing.md:41]: Those will be added in a later PR.

File: format_spec.md (hunk 3 above)
Comment (stairman06) [reply format_spec.md]: If we support a standard API or anything like that we can change the spec to allow it. But right now there is nothing like that supported.

**Comment by stairman06** [comment]
This is not perfect, but it's good for now.

**Pull request merged**